hamilton/caching/adapter.py

```diff
@@ -0,0 +1,1335 @@
+            node_name=self.node_name,
+            task_id=self.task_id,
+            actor=self.actor,
+            event_type=self.event_type.value,
+            msg=self.msg,
+            value=str(self.value) if self.value else self.value,
+        )
+
+
+# TODO we could add a "driver-level" kwarg to specify the cache format (e.g., parquet, JSON, etc.)
+class SmartCacheAdapter(
+    BaseDoNodeExecute, BasePreGraphExecute, BasePostNodeExecute, BasePreNodeExecute
+):
+    """Adapter enabling Hamilton's caching feature through ``Builder.with_cache()``
+
+    .. code-block:: python
+
+        from hamilton import driver
+        import my_dataflow
+
+        dr = (
+            driver.Builder()
+            .with_modules(my_dataflow)
+            .with_cache()
+            .build()
+        )
+
+        # then, you can access the adapter via
+        dr.cache
+
+    """
+
+    def __init__(
+        self,
+        path: Union[str, pathlib.Path] = "hamilton_cache",
+        metadata_store: Optional[MetadataStore] = None,
+        result_store: Optional[ResultStore] = None,
+        recompute: Optional[Union[Literal[True], Collection[str]]] = None,
+        disable: Optional[Union[Literal[True], Collection[str]]] = None,
+        ignore: Optional[Union[Literal[True], Collection[str]]] = None,
+        log_to_file: bool = False,
+        **kwargs,
+    ):
+        """Initialize the cache adapter.
+
+        :param path: path where the cache metadata and results will be stored
+        :param metadata_store: BaseStore handling metadata for the cache adapter
+        :param result_store: BaseStore caching dataflow execution results
+        :param recompute: Set caching behavior to RECOMPUTE for specified node names. If True, apply to all nodes."
+        :param disable: Set caching behavior to DISABLE for specified node names. If True, apply to all nodes.
+        :param ignore: Set caching behavior to IGNORE for specified node names. If True, apply to all nodes."
+        :param log_to_file: If True, append cache event logs as they happen in JSONL format.
+        """
+        self._path = path
+        self.metadata_store = (
+            metadata_store if metadata_store is not None else SQLiteMetadataStore(path=path)
+        )
+        self.result_store = (
+            result_store if result_store is not None else FileResultStore(path=str(path))
+        )
+        self.log_to_file = log_to_file
+
+        if sum([recompute is True, disable is True, ignore is True]) > 1:
+            raise ValueError(
+                "Can only set one of (`recompute`, `disable`, `ignore`) to True. Please pass mutually exclusive sets of node names"
+            )
+        self._recompute = recompute
+        self._disable = disable
+        self._ignore = ignore
+
+        # attributes populated at execution time
+        self.run_ids: List[str] = []
+        self._fn_graphs: Dict[str, FunctionGraph] = {}  # {run_id: graph}
```

Comment:
I thoroughly reviewed the code last week and intentionally set the `_` prefix for internal and external attributes and methods. 

Is there a specific question or concern?

Reply:
My point was that variables such as `behaviors` didn't feel public. Might be wrong though? And `run_ids`?

Reply:
Those are meant for debugging, but I would consider that public still? A common pattern for all of those would be 

```python
# assuming multiple executions occured

run_id = dr.cache.run_ids[3]
cache_key = dr.cache.cache_keys[run_id]["trained_model"]
dr.cache.result_store.get(cache_key)
```

To the contrary, `._disable` isn't "resolved" yet (it might be a list of node names or `True`), so it's internal. For `_fn_graphs`, I felt that this mechanism was a bit of a hack and everything you need should be available through `.context_keys`, `.code_versions`, etc. The access to function graphs would probably change in 2.0

Reply:
So I think better practice is to have them be underscores then expose them with `@property`. This way you can add documentation, and have some semblance of a contract.

Otherwise, exposing instance vars is common, but not great practice. Can just have a TODO here, listing the public vars out if you don't want to make the change.